MAINT: version pins/prep for 1.15.0rc1

tylerjereddy opened the pull request:
Important: CI is skipped completely for now because this will not work--it mostly demonstrates what I'd *like* to be able to do, but that apparently cannot be done because of the limited hooks available in https://packaging.python.org/en/latest/specifications/dependency-specifiers/#dependency-specifiers. I'm pretty sure our only option at the moment for free-threaded CPython `3.13` wheels will be to treat `pyproject.toml` more or less like free-threaded CPython doesn't exist, and disable build isolation for `cibuildwheel` (which is basically what we do on `main`). 

However, if you wanted to do a more formally-correct build-isolation-enabled CPython `3.13` free-threaded wheel build it looks like you'd be in trouble at the moment?

* Update dependency version ranges for build/runtime as we prepare for the SciPy `1.15.0rc1` pre-release, paying special attention to the complications related to free-threaded CPython `3.13` support.

* In addition to the differences caused by free-threading support, there are also some version handling changes that mean this update differs from the previous one performed in the last release cycle at gh-20831.

[skip ci] [ci skip] [skip cirrus] [skip circle]

TODO:

- [ ] regular CI passing
- [ ] wheels CI passing

tylerjereddy committed:
WIP, MAINT: version pins/prep for 1.15.0rc1

* Update dependency version ranges for build/runtime
as we prepare for the SciPy `1.15.0rc1` pre-release,
paying special attention to the complications related
to free-threaded CPython `3.13` support.

* In addition to the differences caused by free-threading
support, there are also some version handling changes
that mean this update differs from the previous one
performed in the last release cycle at gh-20831.

[skip ci] [ci skip] [skip cirrus] [skip circle]

tylerjereddy commented on pyproject.toml:
```diff
@@ -17,18 +17,33 @@
+    "meson-python>=0.15.0,<0.20.0",
+    # we need at least Cython 3.1.0a1 for free-threaded CPython
+    # 3.13 wheel builds
+    "Cython>=3.1.0a1,<3.2.0; python_version=='3.13t'",        # when updating version, also update check in meson.build
```
`3.13t` doesn't exist, so this is invalid, and there's no valid hook via https://packaging.python.org/en/latest/specifications/dependency-specifiers/#dependency-specifiers for accessing free-threaded status that I'm aware of...

rgommers replied:
Correct. I suggest converting the "needs >=3.1.0a1 for free-threading" to a code comment here.

h-vetinari replied:
> Correct. I suggest converting the "needs >=3.1.0a1 for free-threading" to a code comment here.

Not that it changes the lack of specifier for 3.13t, but some good news about cython 3.1: https://github.com/cython/cython/issues/6536

rgommers replied:
There's only Discourse threads (IIRC there was another one). I don't feel like reopening that discussion though, since (a) free-threading as a non-default build is supposed to be a temporary state so no static metadata is defensible, and (b) there are way larger holes in specifiers, like not being able to distinguish between 32-bit and 64-bit Python.

rgommers replied:
yay for Cython 3.1 soon 🤞🏼

tylerjereddy commented on pyproject.toml:
```diff
@@ -17,18 +17,33 @@
+    "Cython>=3.1.0a1,<3.2.0; python_version=='3.13t'",        # when updating version, also update check in meson.build
+    # for other CPython versions, the regular pre-emptive
+    # Cython version bounds policy applies
+    "Cython>=3.0.8,<3.1.0; python_version!='3.13t'",
```
same here... I think we'll just have to do the usual thing and disable build isolation, but it is not a very nice solution--having access to `sysconfig.get_config_var("Py_GIL_DISABLED")` might do the track, but don't think we have access to it?

rgommers replied:
No we don't. There is no way to check for free-threading in `pyproject.toml`.

tylerjereddy commented on pyproject.toml:
```diff
@@ -17,18 +17,33 @@
+    # The upper bound on pythran is pre-emptive only; 0.17.0
+    # is released/working at time of writing, and 0.18.x may
+    # have desirable free-threaded CPython 3.13 support
+    "pythran>=0.14.0,<0.19.0",
```
given that the rest of the `pyproject.toml` is likely going to have to "ignore" free-threaded CPython, not sure the extension here makes much sense..

rgommers replied:
Agreed. I'm fine with <0.18.0 here

tylerjereddy commented on pyproject.toml:
```diff
@@ -17,18 +17,33 @@
+    "numpy>=2.0.0,<2.5; python_version!='3.13t'",
+    # free-threaded CPython 3.13 support requires more
+    # recent NumPy release at build time
+    "numpy>=2.1.3,<2.5; python_version=='3.13t'",
```
Again, these won't work... probably just the usual one above and purge the Python version checks + disable build isolation for free-threaded..

rgommers replied:
Same, changing to a code comment would be good.

tylerjereddy commented on pyproject.toml:
```diff
@@ -46,7 +61,10 @@ maintainers = [
+    # free-threaded CPython 3.13 support requires more
+    # recent NumPy release at runtime
+    "numpy>=2.1.3,<2.5; python_version=='3.13t'",
+    "numpy>=1.23.5,<2.5; python_version!='3.13t'",
```
This also doesn't work of course, probably just do the usual. At least for this runtime case, I believe could catch the problem of lowest NumPy req in `__init__.py` for free-threaded if we wanted though?

tylerjereddy commented on scipy/__init__.py:
```diff
@@ -67,7 +67,7 @@
 # In maintenance branch, change to np_maxversion N+3 if numpy is at N
 np_minversion = '1.23.5'
-np_maxversion = '9.9.99'
+np_maxversion = '2.5.0'
```
We could conditionally bump `np_minversion` to `2.1.3`  if `sysconfig.get_config_var("Py_GIL_DISABLED")` is `1` perhaps? Or maybe something like that--we should have more programmatic freedom to do such a detection here vs. in `pyproject.toml` I'd think.

rgommers reviewed (commented):
Thanks Tyler. Added some comments.

tylerjereddy committed:
MAINT, BLD: PR 22024 revisions

* Apply suggestions from code review--attempt to use
`pyproject.toml` version bounds that actually work,
which mostly means ignore free-threaded CPython
altogether.

tylerjereddy commented:
Ok, I've revised this back to something closer to what we usually do, with my complaining about lack of free-threaded CPython detection in `pyproject.toml` out of the way.

I'll let the regular CI flush through now as an initial check--I've added a checklist above that includes running the full wheel builds here as well--I usually prefer to do that in the version bound setting PR to avoid any surprises during the release process proper.

rgommers commented on pyproject.toml:
```diff
@@ -17,18 +17,28 @@
     # redistributors can do this by installing the numpy version they like and
     # disabling build isolation.
-    "numpy>=2.0.0rc1",
+    # NOTE: need numpy>=2.1.3 from free-threaded CPython 3.13 support
```
minor typo: "from" -> "for"

rgommers reviewed (approved):
Failures are unrelated, so this looks good now.

tylerjereddy committed:
MAINT: PR 22024 revisions [wheel build]

* Fix a typo in `pyproject.toml` found by reviewer,
and initiate a test of the full wheel build matrix.

[wheel build]

tylerjereddy commented:
I pushed in a fix for the typo and initiated a full wheel build at the same time--I'd like to see where we stand with wheel builds for the pre-release now.

There are a few backports to consider already re: unrelated CI failures too, but I think most of them are minor enough that it is "ok" to do the assessment of the wheel build situation now and then decide if I should backport them before RC1.

tylerjereddy commented:
Apart from failures that can be fixed from known backports, here are some other failures showing up now with wheel builds running:

1. Cirrus is timing out at 75-85 % testsuite run time for the second Python version in each wheel build matrix entry
2. Windows free-threading test issue: https://github.com/scipy/scipy/issues/22039
3. a `pytest-fail-slow` timing failure that may be flaky and was reported by Ralf last month at https://github.com/scipy/scipy/issues/20806#issuecomment-2466918527

A few wheel builds are still running--I'll check on them in a bit. For point 1, we can probably increase the timeout limits on the release branch--we've done that before. For point 2, not sure yet.

rgommers commented:
> For point 1, we can probably increase the timeout limits on the release branch--we've done that before.

Yes, not unexpected - the builds go heavier I think, especially with HiGHS being more bulky now.

> Windows free-threading test issu

I'll have a look.

> `pytest-fail-slow` timing failure

We shouldn't install `pytest-fail-slow` in wheel builds at all.

tylerjereddy commented:
I only pushed a typo fix after Ralf's approval, so let's get this in and I'll start pushing on a round of backports and fixes for remaining CI issues.
